### PR TITLE
Edge case bugs when using `aggregate_downsample` with custom binning

### PR DESCRIPTION
**Pull request by barentsen**
In Astropy v5.0.0, the `astropy.timeseries.aggregate_downsample` function was modified to allow the `time_bin_start` and `time_bin_size` parameters to be arrays (cf. #11266).  This is an excellent feature which is working very well overall (thank you @aaryapatil +  @dhomeier + others!!)

While using this new feature, I encountered a few minor bugs related to edge cases.  This PR attempts to report+fix these issues and adds regression tests.

## Issue 1: `n_bins` needs to be ignored if `time_bin_start` is an array

A confusing `IndexError` is raised if the input for `time_bin_start` is an array and `n_bins` is not equal to its length.  This PR proposes to ignore the value of `n_bins` if `time_bin_start` is an array (i.e. the length of the array takes precedence over the value passed via `n_bins`).  This would make the behavior consistent with the constructor of [BinnedTimeSeries](https://docs.astropy.org/en/stable/api/astropy.timeseries.BinnedTimeSeries.html#binnedtimeseries).

### Example before this PR:
```python
>>> from astropy.time import Time
>>> from astropy.timeseries import TimeSeries, aggregate_downsample
>>>
>>> t = Time(['2016-03-22T12:30:31', '2016-03-22T12:30:32', '2016-03-22T12:30:33'])
>>> ts = TimeSeries(time=t, data=[[1, 1, 1]], names=['a'])
>>> aggregate_downsample(ts, time_bin_start=t, n_bins=2)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/gb/dev/astropy/astropy/timeseries/downsample.py", line 204, in aggregate_downsample
    data[unique_indices] = reduceat(values, groups, aggregate_func)
  File "/Users/gb/.pyenv/versions/astropy/lib/python3.9/site-packages/numpy/ma/core.py", line 3385, in __setitem__
    _data[indx] = dval
IndexError: index 2 is out of bounds for axis 0 with size 2
```

### After this PR:
```python
>>> aggregate_downsample(ts, time_bin_start=t, n_bins=2)
<BinnedTimeSeries length=3>
     time_bin_start       time_bin_size      a
                                s
          Time               float64       int64
----------------------- ------------------ -----
2016-03-22T12:30:31.000 1.0000000000065512     1
2016-03-22T12:30:32.000 0.9999999999969589     1
2016-03-22T12:30:33.000                0.0     1
```

## Issue 2: bins are not allowed to fall beyond the data

Currently (AstroPy v5.0.0), an `IndexError` is raised if all the requested bins fall beyond the time span covered by the data.  This PR fixes that issue.

### Before:

```python
>>> ts = TimeSeries(time=t[0:2], data=[[1, 1]], names=['a'])
>>> aggregate_downsample(ts, time_bin_start=t[2:])
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/gb/dev/astropy/astropy/timeseries/downsample.py", line 174, in aggregate_downsample
    indices[indices_start] = np.arange(len(indices_start))
IndexError: index 0 is out of bounds for axis 0 with size 0
```

### After:

```python
>>> aggregate_downsample(ts, time_bin_start=t[2:])
<BinnedTimeSeries length=1>
     time_bin_start     time_bin_size   a
                              s
          Time             float64    int64
----------------------- ------------- -----
2016-03-22T12:30:33.000           0.0    --
```

## Issue 3: `time_bin_size` is sometimes negative!

Currently, the default value for `time_bin_size` can be negative, which is confusing. 

### Before:
```python
>>> ts = TimeSeries(time=t[[0]], data=[[1]], names=['a'])
>>> aggregate_downsample(ts, time_bin_start=t).time_bin_size
<Quantity [ 1.,  1., -2.] s>
```

### After:
```python
>>> aggregate_downsample(ts, time_bin_start=t).time_bin_size
<Quantity [1., 1., 0.] s>
```

(Note: I believe 1 or 2 similar bugs may still be lingering around.)

### CODE REVIEW

**Comment by hamogu** [comment]
I think this looks great. I'll leave it for a few days if @aaryapatil or @dhomeier want to have a look, too; if we don't hear from them we should approve and merge!

**Comment by dhomeier** [comment]
Thanks for tracking these bugs @barentsen and apologies I haven't had much time for reviewing yet!
The fixes are working as intended; I mainly have concerns about silently fixing inadequate output especially in the first two cases. The first reminds me of a similar discussion on `lightcurve.bin()` about establishing a clear priority between the `time_bin_start`, `n_bins`, `nbins` etc. keywords that was never fully completed, but it is in fact perhaps better addressed here. `time_bin_start` taking precedence here is correct, but I am wondering if a warning should be given about ignoring the `n_bins` input rather than silently discarding it. Otherwise at least the docstring should state that this will conflict with non-scalar `time_bin_start` and be ignored.
For the second case of all empty or masked bins the case for a warning is perhaps weaker, but I remember stumbling over empty binned lightkurves where accidentally the bins were set all outside the time range.
The third case actually goes a bit deeper; while somewhat weird results should perhaps be expected when binning a single-point series into multiple bins, the actual values are plain wrong:
```python
>>> ts = TimeSeries(time=t[[0]], data=[[1]], names=['a'])
>>> aggregate_downsample(ts, time_bin_start=t)
<BinnedTimeSeries length=3>
     time_bin_start       time_bin_size      a  
                                s               
          Time               float64       int64
----------------------- ------------------ -----
2016-03-22T12:30:31.000 1.0000000000065512    --
2016-03-22T12:30:32.000 0.9999999999969589    --
2016-03-22T12:30:33.000                0.0     1
```
– correct bin size with your patch, but ending up in the wrong bin both before and after – which probably was responsible for the negative size as well. I guess this is due to `time_bin_end[-1]` being set to `ts.time[-1]`, but this needs further investigation, especially why the bin for `time_bin_start = t[0]` is somehow ignored or lost here.

**Comment by barentsen** [comment]
> Otherwise at least the docstring should state that this will conflict with non-scalar time_bin_start and be ignored.

As part of this PR, I added the following sentence to the docstring for `n_bins`:
```
If ``time_bin_start`` is an iterable, this parameter will be ignored.
```
I'm not personally a fan of issuing warnings for behavior that is documented by the API.  However, I can easily be convinced to add a warning if we think there is a real potential for scientific error here.

> The third case actually goes a bit deeper; while somewhat weird results should perhaps be expected when binning a single-point series into multiple bins, the actual values are plain wrong:

Yes, this PR does not fully resolve the weird results related to "Issue 3" yet.  I believe the logic in lines 167-177 are to blame for the weird results, but I've gotten stuck trying to figure out how the logic can be improved. I'm referring to this snippet:

https://github.com/astropy/astropy/blob/e97ebc78f7e17fbcbaff47d6ea52ef4130f30ea9/astropy/timeseries/downsample.py#L167-L177

**Comment by dhomeier** [comment]
> As part of this PR, I added the following sentence to the docstring for `n_bins`:
> 
> ```
> If ``time_bin_start`` is an iterable, this parameter will be ignored.
> ```
> 
> I'm not personally a fan of issuing warnings for behavior that is documented by the API. However, I can easily be convinced to add a warning if we think there is a real potential for scientific error here.
> 
Ah, sorry, I was looking at the wrong file for the update in the docs. I don't think there is real danger for an error, as the length of the binned series will simply not match the number requested by `n_bins`; my main concern was probably that it somewhat deviates from the handling of conflicting `time_bin_size`, `time_bin_end` etc. in `BinnedTimeSeries` which generally raises `TypeError` right away. But since part of downsample's functionality is to intercept these conflicting settings, I am comfortable with keeping it documented as is.

> > The third case actually goes a bit deeper; while somewhat weird results should perhaps be expected when binning a single-point series into multiple bins, the actual values are plain wrong:
> 
> Yes, this PR does not fully resolve the weird results related to "Issue 3" yet. I believe the logic in lines 167-177 are to blame for the weird results, but I've gotten stuck trying to figure out how the logic can be improved. I'm referring to this snippet:
> 
> https://github.com/astropy/astropy/blob/e97ebc78f7e17fbcbaff47d6ea52ef4130f30ea9/astropy/timeseries/downsample.py#L167-L177

Um, yes! What's happening there is that a single-valued array `indices = np.array([0])` is found from the ends, then the starts find all-zero `indices_start = [0, 0, 0]`, thus setting `indices[[0, 0, 0]] = np.arange(3)`, which is keeping the last value in the range!
I suppose this can be mended by using `np.arange(min(len(indices_start), len(indices)))` instead, but not sure if that will really cover all eventualities.
Actually
```python
    if np.all(bin_start[1:] >= bin_end[:-1]):
        indices_start = np.searchsorted(subset.time, bin_start[bin_start <= ts_sorted.time[-1]])
        indices[indices_start] = np.arange(len(indices_start))
```
makes a bit more sense, and even works when setting empty slices to `np.arange(0)`.

**Comment by aaryapatil** [comment]
Apologies for the delay in reviewing this PR. Thank you for detecting these bugs and solving them! There were lots of cases to handle here, so it is great to see these changes.

I agree with @dhomeier that silently fixing issues with parameters is not ideal, but also agree that the first two issues do not warrant a ``TypeError``. In ``aggregate_downsample``, ``TypeError``s are only being raised in cases where the input is inadequate. In cases where there are conflicting parameters, docstrings should be adequate to guide the use; unless there is something "bad" happening behind the scenes, e.g. the AstropyUserWarning for avoiding double counting with overlapping bins -- this is how I implemented the changes.

Issue 3 is definitely concerning. Hopefully the above suggested fix works!

Other than that, this looks great 😄

**Comment by hamogu** [comment]
So, what do we do? I think we all agree that this is an improvement over what's in the code now; we also seem to agree that no more errors or warnings are needed. The question is: Do we merge this as it's now and leave future improvements for another PR or would you be willing to invest the time and look into @dhomeier suggestion in the last few lines of this comment https://github.com/astropy/astropy/pull/12527#issuecomment-984019182 to address maybe a few more edge cases (negative bin sizes) to fix that here while we are at it?

File: astropy/timeseries/downsample.py
Comment (dhomeier) [inline astropy/timeseries/downsample.py:178]:
```suggestion
        indices_start = np.searchsorted(subset.time, bin_start[bin_start <= ts_sorted.time[-1]])
```
Putting this forward now from https://github.com/astropy/astropy/pull/12527#issuecomment-984019182 as it fixes at least the known instances of Issue 3.

**Review by dhomeier — CHANGES_REQUESTED** [review]
I agree that this is fixing enough issues to get in as an improvement (preferably + the suggested fix for Issue 3). Test coverage for that would be nice, but since 5.0.1 is now tentatively scheduled for 7 January, merging this in time should have priority.

File: astropy/timeseries/tests/test_downsample.py
Comment (dhomeier) [inline astropy/timeseries/tests/test_downsample.py:191]:
```suggestion
@pytest.mark.parametrize("time, time_bin_start, time_bin_end",
                        [(INPUT_TIME[:2], INPUT_TIME[2:], None),
                         (INPUT_TIME[3:], INPUT_TIME[:2], INPUT_TIME[1:3]),
                         (INPUT_TIME[[0]], INPUT_TIME[:2], None),
                         (INPUT_TIME[[0]], INPUT_TIME[::2], None)])
def test_downsample_edge_cases(time, time_bin_start, time_bin_end):
    """Regression test for #12527: allow downsampling even if all bins fall
    before or beyond the time span of the data."""

    ts = TimeSeries(time=time, data=[np.ones(len(time))], names=['a'])
    down = aggregate_downsample(ts, time_bin_start=time_bin_start, time_bin_end=time_bin_end)
    assert len(down) == len(time_bin_start)
    assert all(down['time_bin_size'] >= 0)  # bin lengths shall never be negative
    if ts.time.min() < time_bin_start[0] or time_bin_end is not None:
        assert down['a'].mask.all()        # all bins placed *beyond* the time span of the data
    elif ts.time.min() < time_bin_start[1]:
        assert down['a'][0] == ts['a'][0]  # single-valued time series falls in *first* bin
```
This adds two simple tests for case 3.

**Comment by barentsen** [comment]
Thanks @dhomeier!  I merged the tests you suggested for issue 3, verified that the tests failed, then merged the fix you suggested, and verified that the tests are passing now.

I won't have a lot of time in the immediate future to hunt down additional edge cases.  I believe this PR fixes the most obvious edge cases, and would be beneficial to include in v5.0.1.

**Comment by hamogu** [comment]
@barentsen Can you rebase on main? That should fix the readthedocs failure. Let me know if you don't know how or don't have time and I or @dhomeier can do that for you. I think this looks otherwise good (will check line by line again right now) but we can't merge it until all required tests pass.

**Comment by barentsen** [comment]
I will rebase in the next hour!

**Comment by barentsen** [comment]
@hamogu Rebase complete.  All tests passing! 🎉

**Review by hamogu — APPROVED** [review]
Looks good to me. @dhomeier had requested changes, so I hold of on merging right now to give him a chance to look again, but as far as I'm concerned, this can go in any time.

Thanks for spotting and fixing those bugs!

**Review by dhomeier — APPROVED** [review]

**Comment by dhomeier** [comment]
Thanks a lot for getting this in in time @barentsen; yes, this is fixing all bugs known at this point, and if any more should show up, they can be addressed later!

**Pull request merged**
